### ISSUE TITLE
hbool_t on Windows and accessing HDF5 config header

### ISSUE THREAD

**Issue by aldanor** [aldanor/hdf5-rust#28]
So, we currently have `hbool_t` mapped as `c_uint` which turns out to be wrong because on Linux / OSX it's an alias of C `bool` which is 1-byte. As a result, some struct layouts get completely jumbled (which is pretty hard to check other than observing stuff failing in obscure ways).

Once I've changed it to 1 byte, it worked out perfectly on Unix, but some Windows tests started failing on AppVeyor (thanks god we have some kind of an actual test suite...). Took me a while to narrow down the issue (I had to install Windows with MSVS in a virtual machine...), turns out to be this:

```c
#ifdef H5_HAVE_STDBOOL_H
  #include <stdbool.h>
#else /* H5_HAVE_STDBOOL_H */
  #ifndef __cplusplus
    #if defined(H5_SIZEOF_BOOL) && (H5_SIZEOF_BOOL != 0)
      #define bool    _Bool
    #else
      #define bool    unsigned int
    #endif
    #define true    1
    #define false   0
  #endif /* __cplusplus */
#endif /* H5_HAVE_STDBOOL_H */
typedef bool hbool_t;
```

Now, the way it works with HDF5 builds, all those `H5_*` constants are auto-generated by the CMake process at build time, and dumped into a special header, `H5pubconf.h`, which is distributed along with the binaries and the source headers. It appears that whichever machine they were building this on, `stdbool.h` header was not accessible (which is somewhat weird because it has been available in MSVS since v2013). Apparently the CMake process has been [fixed upstream](https://github.com/live-clones/hdf5/commit/6f6d0921c0592c202bae15dab1b05c921aef0043) about two weeks ago, so it will now use assume C99, and thus use proper bool type – however, we can't make use of it since we have to support all the older versions of the library.

We could just say, of course,

```rust
#[cfg(target_env = "msvc")]
pub type hbool_t = c_uint;
#[cfg(not(target_env = "msvc"))]
pub type hbool_t = u8;
```

... but this is generally wrong, dodgy and fragile. Because, for instance, 3 years ago it looked like this:

```c
typedef unsigned int hbool_t;
```

Unfortunately, all those `H5_*` constants are not linked in into the library itself, so we can't access them the same way we extract library version.

... which leads to another point -- having those constants available as compile-time things, or at least some of them, like `H5_HAVE_PARALLEL`, `H5_THREADSAFE`, `H5_HAVE_DIRECT` would allow us to provide proper conditionally-compiled bindings for those cases (MPI support, thread-safe build and direct driver support, respectively).

Summarizing:

- Pros:
  - we gain access to auto-generated information about how the library was built
  - we can provide conditional bindings for MPI stuff, direct driver support, thread-safe builds etc
  - types like `hbool_t` will be properly bound
- Cons:
  - we have to assume `include/H5pubconf.h` always exists, we won't be able to build the crate without it (unless we make some assumptions which I'd rather not make at build time). This seems reasonable because the library is always distributed with the headers. Which means we'll have to search for the header kind of the same way we search for the library files; which might mean hard-coding library directory structure layout on windows and linux/osx.
  - we will either have to compile a C program into a crate (e.g. via using cc-rs; this implies access to a C compiler), or parse the header ourselves manually.

Options available:

1. Compile a simple C program that dumps all this stuff (or provides it as a crate) via [`cc-rs`](https://github.com/alexcrichton/cc-rs). Now this assumes existence of a C compiler on the current platform, I'm not sure but that sounds like an excessively harsh requirement.
2. Parse this C header manually with regex which shouldn't be too bad in theory  since it's just a bunch of `#define`s and not much else). We could probably even auto-generate a Rust module out of it with all those constants as `pub const` variables. We would obviously only be able to parse definitions not containing any expressions (that is, literals only).

**Comment by aldanor** [comment]
This would actually make a good first issue if anyone wants to contribute since it's independent from the rest of the codebase and doesn't require deep knowledge of the library - if we go for the second option above, it would just require parsing some text file with regex and outputting another text file.

**Comment by Enet4** [comment]
I might not have the full picture of this at the moment, but would it be feasible for the `-sys` crate to look for the header file and `bindgen` that at build-time? I feel that this would be preferable over regexing the header.

**Comment by aldanor** [comment]
@Enet4 I'm not sure how it would help because some of these may or may not be defined (intended to be used with `#ifdef`).

For example, if there is `#define H5_HAVE_DIRECT 1`, it means that you have access to the direct driver, and all the companion methods in the library. However, if it's disabled (or if you're using an ancient version of the library that doesn't have it), you won't see anything.

So if you were to run bindgen, you would either see a `pub const H5_HAVE_DIRECT: c_int = 1;` or you would see... nothing.

**Comment by aldanor** [comment]
I've also discovered that `hbool_t` may be 4 bytes on Linux, too! Holy cow. Here's the commit:

https://github.com/live-clones/hdf5/commit/2094d8646199996f655bb5f56f142216bc2abfcf

**Comment by aldanor** [comment]
> I think we do not have to worry about the possibility of H5's boolean type being defined as an unsigned int.

Not correct. Apparently all HDF5 1.8 would have 4 byte bool if I understand it correctly: https://github.com/live-clones/hdf5/commit/2094d8646199996f655bb5f56f142216bc2abfcf

I've just printed `sizeof(hbool_t)` in a test C program on a modern Linux with g++-5 and HDF5 1.8.18 and it seems to be 4.

**Comment by aldanor** [comment]
Just to note, the problem is, in the commit linked above they fixed *most* of the existing struct fields from `hbool_t` to `unsigned` to retain backwards-compatibility... but not all of them. Here's one: https://github.com/live-clones/hdf5/blob/develop/src/H5ACpublic.h#L452

**Comment by Enet4** [comment]
Hmm. This isn't fun any more. :disappointed: Can we still try to bindgen everything into place, or will this still bring us compatibility issues?

**Comment by aldanor** [comment]
> Can we still try to bindgen everything into place, or will this still bring us compatibility issues?

What kind of "compatibility issues"? We can generate proper bindings that will be correct with a very high probability just for *this particular version* of HDF5 that you're currently using. This is already true in many other places because HDF5 library has been changing over time, sometimes in a non-backwards compatible way (like it is with hbool_t, but there are other similar issues).

I think we can manually bindgen the `H5pubconf.h` file, using macro callback functionality, extract the macros that interest us (including library version, to match it with `H5get_libversion()` just to ensure we're not doing anything stupid), and pass them on as `--cfg` to all downstream crates (`libhdf5-sys` and `hdf5-rs`), similar to how it's currently done with library versions. We could then do something like this:

```rust
#[cfg(hdf5_sizeof_hbool_t = "1")]
pub type hbool_t = u8;
#[cfg(hdf5_sizeof_hbool_t = "4")]
pub type hbool_t = u32;

#[cfg(hdf5_have_direct)]
mod direct {
    // core driver API
}
#[cfg(hdf5_have_direct)]
pub use self::direct::*;
```

**Comment by aldanor** [comment]
@Enet4 this is now fixed in #29 (I think).
